src/FsAutoComplete.Core/Utils.fs

```diff
@@ -736,3 +736,24 @@ type Debounce<'a>(timeout, fn) =
 
   /// Calls the function, after debouncing has been applied.
   member __.Bounce(arg) = mailbox.Post(arg)
+
+let filePathContents (path: string) =
+    Path.GetDirectoryName(path)
+        .Split(Path.DirectorySeparatorChar, System.StringSplitOptions.RemoveEmptyEntries)
+    |> Array.toList
+
+let createDirectoryIfMissing projPath dirs =
+    let rec create dirs =
+        match dirs with
+        | [] -> ()
+        | dir :: dirs ->
+            let dir' = Path.Combine(projPath, dir)
+            if not <| Directory.Exists dir' then
+                Directory.CreateDirectory dir' |> ignore
```

Comment:
[`Directory.CreateDirectory(dir: string)`](https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.createdirectory?view=net-6.0) creates all necessary intermediate directories

> Creates all directories and subdirectories in the specified path unless they already exist.

So you should be able to simplify this function (and its inputs) quite a lot!

Reply:
Hm, not sure why I though that was an issue. The bug I mentioned still exists.

Reply:
Where is FsProjAddFile used?

Reply:
All of these LSP endpoints for project file manipulation are mostly used by ionide to implement right-click menu items for the solution explorer. This is implemented [here](https://github.com/ionide/ionide-vscode-fsharp/blob/460eb38b85fef4cf63c5ef330e636051daef307a/src/Core/FsProjEdit.fs#L18-L35) if you want to check out the implementation.

Reply:
We've got them implemented in the LSP server side of things so that, say, ionide-vim or the emacs mode could easily have the same feature set

Reply:
Sorry, I meant in the UI. I can't figure out how to test it. I thought it might be the "add current file to project" command but that fails regardless ("n is not a function" - not even sure it's hitting LSP).

Reply:
That doesn't sound good.  The command palette version of the 'add file to project' command is defined in this file, but all of the solution file commands are [here](https://github.com/ionide/ionide-vscode-fsharp/blob/460eb38b85fef4cf63c5ef330e636051daef307a/src/Components/SolutionExplorer.fs#L770-L874), they call into the functions I linked above.  Does that help find what's happening?